pkg/controller/snapshotter/snapshotter.go

```diff
@@ -273,38 +286,72 @@ func (vs *volumeSnapshotter) getSnapshotCreateFunc(snapshotName string, snapshot
 			Metadata: metav1.ObjectMeta{
 				Name: snapName,
 			},
-			Spec: crdv1.VolumeSnapshotDataSpec{
-				VolumeSnapshotRef: &v1.ObjectReference{
-					Kind: "VolumeSnapshot",
-					Name: snapshotName,
-				},
-				PersistentVolumeRef: &v1.ObjectReference{
-					Kind: "PersistentVolume",
-					Name: pvName,
-				},
-				VolumeSnapshotDataSource: *snapshotDataSource,
-			},
 			Status: crdv1.VolumeSnapshotDataStatus{
 				Conditions: []crdv1.VolumeSnapshotDataCondition{
 					readyCondition,
 				},
 			},
 		}
+
 		var result crdv1.VolumeSnapshotData
-		err = vs.restClient.Post().
-			Resource(crdv1.VolumeSnapshotDataResourcePlural).
-			Namespace(v1.NamespaceDefault).
-			Body(snapshotData).
-			Do().Into(&result)
+		for i := 0; i < createVolumeSnapshotDataRetryCount; i++ {
+			err = vs.restClient.Post().
+				Resource(crdv1.VolumeSnapshotDataResourcePlural).
+				Namespace(v1.NamespaceDefault).
+				Body(snapshotData).
+				Do().Into(&result)
+			if err != nil {
+				// Re-Try it as errors writing to the API server are common
+				glog.Infof("Error creating the VolumeSnapshotData %s: %v. Retrying...", snapshotName, err)
+				time.Sleep(createVolumeSnapshotDataInterval)
+			} else {
+				break
+			}
+		}
 
 		if err != nil {
-			//FIXME: Errors writing to the API server are common: this needs to be re-tried
 			glog.Errorf("Error creating the VolumeSnapshotData %s: %v", snapshotName, err)
+			// Don't proceed to create snapshot using the plugin due to error creating
+			// VolumeSnapshotData
+			return fmt.Errorf("Failed to create the VolumeSnapshotData %s for snapshot %s", snapName, snapshotName)
 		}
+
+		// Create a snapshot on the storage backend using the plugin
+		snapshotDataSource, err := vs.takeSnapshot(pv)
```

Comment:
Currently the workflow is: 
- take a snapshot 
- create `VolumeSnapshotData` with ready condition
- update `VolumeSnapshot`

With this changes, the workflow becomes: 
- create `VolumeSnapshotData`, 
- take a snapshot
- update status in `VolumeSnapshotData` 
- update `VolumeSnapshot`.

None of these is controller crash safe. We probably better do:
- create snapshot metadata info to help to uniquely identify a snapshot
- if `VolumeSnapshot` binds to `VolumeSnapshotData` that has the same metadata info, then don't take more snapshot
otherwise:
- create `VolumeSnapshotData`
- use this metadata info in `VolumeSnapshotData`
- update `VolumeSnapshot` with `VolumeSnapshotData`
- take a snapshot with this metadata info
- update status in `VolumeSnapshotData` 

@jingxu97 @tsmetana